orttraining/orttraining/python/training/_ortmodule_training_manager.py

```diff
@@ -183,8 +188,28 @@ def _create_execution_agent(self):
         """Creates a TrainingAgent that can run the forward and backward graph on the training model"""
 
         session_options, providers, provider_options = self._get_session_config()
+
+        fw_feed_names = []
+        for idx, value_info in enumerate(self._optimized_onnx_model.graph.input):
+            fw_feed_names.append(value_info.name)
+        fw_outputs_device_info = []
+        for idx in range(len(self._graph_info.user_output_names)):
+            fw_outputs_device_info.append(C.OrtDevice(self._get_ort_device_type(self._device.type),
+            C.OrtDevice.default_memory(), U.get_device_index(self._device)))
+
+        bw_fetches_names = []
+        for idx, value_info in enumerate(self._optimized_onnx_model.graph.output):
+            bw_fetches_names.append(value_info.name)
+        bw_outputs_device_info = []
+        for idx in range(len(bw_fetches_names)):
+            bw_outputs_device_info.append(C.OrtDevice(self._get_ort_device_type(self._device.type),
+            C.OrtDevice.default_memory(), U.get_device_index(self._device)))
```

Comment:
Possible optimization: it doesn't look like we need to construct this list of `len(bw_fetches_names)` identical objects explicitly. Perhaps we could optimize whatever interface takes this argument to accept a single element and "broadcast" it for all the outputs?

(Edit: I guess this only happens on the first run, so maybe not worth the effort.)

Reply:
I did also think of this device info (since really all these values will be the same) and didn't bother optimizing since it only happens on the first run. 

---
In reply to: [613576856](https://github.com/microsoft/onnxruntime/pull/7324#discussion_r613576856) [](ancestors = 613576856)

Reply:
Naive question, haven't complete reading the PR here... 
I suppose bw_outputs_device_info is trying to solve the problem that different bw_outputs needs to live in different device? 
Do have such a scenario?